ambry-store/src/main/java/com.github.ambry.store/DiskSpaceAllocator.java

```diff
@@ -0,0 +1,374 @@
+   *     certain segment size, the difference is added into the overall requirements map. This ensures that the total
+   *     (allocated and pooled) swap segment count equals the required count.
+   *   </li>
+   *   <li>
+   *     Unneeded segments are deleted. These are segments that were present in the pool on disk, but are not needed
+   *     according to the overall requirements.
+   *   </li>
+   *   <li>
+   *     Any additional required segments are added. For each segment size, additional files will be preallocated to
+   *     until the number of files at that size match the overall requirements map. If fallocate is supported on the
+   *     system, and the call fails, pool initialization will fail.
+   *   </li>
+   * </ol>
+   * @param requirementsList a collection of {@link DiskSpaceRequirements}s objects that describe the number and
+   *                         segment size needed by each store.
+   * @throws StoreException if the pool could not be allocated to meet the provided requirements
+   */
+  void initializePool(Collection<DiskSpaceRequirements> requirementsList) throws StoreException {
+    long startTime = System.currentTimeMillis();
+    try {
+      Map<Long, Long> overallRequirements = getOverallRequirements(requirementsList);
+      deleteExtraSegments(overallRequirements);
+      addRequiredSegments(overallRequirements);
+      // TODO fill the disk with additional swap segments
+      poolInitialized = true;
+    } catch (Exception e) {
+      metrics.diskSpaceAllocatorInitFailureCount.inc();
+      throw new StoreException("Exception while initializing DiskSpaceAllocator pool", e,
+          StoreErrorCodes.Initialization_Error);
+    } finally {
+      long elapsedTime = System.currentTimeMillis() - startTime;
+      LOGGER.info("initializePool took {} ms", elapsedTime);
+      metrics.diskSpaceAllocatorStartTimeMs.update(elapsedTime);
+    }
+  }
+
+  /**
+   * Allocate a file, that is, take a file matching the requested size from the reserve directory and move it to the
+   * provided file path
+   * @param destinationFile the file path to move the allocated file to.
+   * @param sizeInBytes the size in bytes of the requested file.
+   * @throws IOException if the file could not be moved to the destination.
+   */
+  void allocate(File destinationFile, long sizeInBytes) throws IOException {
+    if (!poolInitialized) {
+      LOGGER.info("Allocating segment of size {} to {} before pool is fully initialized", sizeInBytes,
+          destinationFile.getAbsolutePath());
+    }
+    File reserveFile = reserveFiles.remove(sizeInBytes);
+    if (reserveFile == null) {
+      LOGGER.info("Segment of size {} not found in pool; attempting to create a new preallocated file", sizeInBytes);
+      metrics.diskSpaceAllocatorSegmentNotFoundCount.inc();
+      reserveFile = createReserveFile(sizeInBytes);
+    }
+    if (destinationFile.exists() || !reserveFile.renameTo(destinationFile)) {
```

Comment:
if destinationFile exists, can we fail earlier (even before creating a new reserve file or fetching one from pool?